common/src/main/java/com/github/khanshoaib3/minecraft_access/features/inventory_controls/GroupGenerator.java

```diff
@@ -467,8 +466,51 @@ else if (screen.getHandler() instanceof HopperScreenHandler)
             foundGroups.add(itemOutputGroup);
         }
 
-        if (unknownGroup.slotItems.size() > 0) {
-            foundGroups.add(unknownGroup);
+        // Adjacency is used here instead of slot.inventory because some mods have non-adjacent slots in the same
+        // inventory and adjacent slots in different inventories
+        List<List<SlotItem>> unknownGroups = new ArrayList<>(unknownSlots.size());
+        for (SlotItem slot : unknownSlots) {
+            unknownGroups.stream()
+                    .filter(group ->
+                            group.stream()
+                                    .anyMatch(groupSlot ->
+                                            groupSlot.x == slot.x && (groupSlot.y == slot.y + 18 || groupSlot.y == slot.y - 18)
+                                            || groupSlot.y == slot.y && (groupSlot.x == slot.x + 18 || groupSlot.x == slot.x - 18)
+                                    )
+                    )
+                    .findFirst()
+                    .or(() -> {
+                        List<SlotItem> group = new ArrayList<>(unknownSlots.size());
+                        unknownGroups.add(group);
+                        return Optional.of(group);
+                    })
+                    .get()
+                    .add(slot);
+        }
+        Map<String, Byte> usedNames = new HashMap<>(unknownGroups.size());
+        for (List<SlotItem> group : unknownGroups) {
+            String groupName;
+            if (group.stream()
+                    .map(slot -> slot.slot.getClass().getSimpleName())
+                    .distinct()
+                    .limit(2)
+                    .count() == 1) {
+                groupName = CaseFormat.UPPER_CAMEL.to(CaseFormat.LOWER_UNDERSCORE, group.get(0).slot.getClass().getSimpleName());
+		// Don't use vanilla obfuscated class names
+                if (groupName.startsWith("class_")) {
+                    groupName = "unknown_group";
+                }
+            } else {
+                groupName = "unknown_group";
+            }
+            if (usedNames.containsKey(groupName)) {
+                byte n = (byte) (usedNames.get(groupName) + 1);
+                usedNames.put(groupName, n);
+                groupName += String.format("_%d", n);
+            } else {
+                usedNames.put(groupName, (byte) 1);
+            }
+            foundGroups.add(new SlotsGroup(groupName, group));
```

Comment:
What's the value of groupName when code goes to here,
for example, number1_number2_number3... ?
This is not readable than unknown group in my view, maybe I misunderstood the meaning of these group naming logic.
And maybe extract this part (in for loop) as a method is good.

Reply:
Essentially, if all the slots in a group have the same class name groupName will be set to that name, if groupName starts with `class_` (vanilla obfuscated classes) or the slots have different classes it will be set to unknown_group. Then if there is already a group with that name it will get a number added to the end.

This approch does have the side effect of not giving the first group a number when there are other groups with the same name but the alternatives are giving every group a number or making the algorithm significantly more complex.

Reply:
Now it make sense to me. Can you add some comments about describing these if-else conditions like what you did in your last comment?

Reply:
TIL: use byte for small numbers. I don't if it's a good usage...

Reply:
> TIL: use byte for small numbers. I don't know if it's a good usage...

There would have to be 128 identically named groups for it to be a problem, I could use a short if that's a concern but I have never seen a mod that has anywhere near 128 slot groups that I can remember and it should wrap around to -128 if that does happen for some reason, realistically the detection of 128 groups is probably itself a bug or a really bad ui and a better mekanism for switching groups may be in order for that many groups. I guess we're not going to be ram limited to the point that a few bytes is significant and it should theoretically have roughly O(n) space complexity so there is no reason not to use a short or even an int or a long.

> Now it make sense to me. Can you add some comments about describing these if-else conditions like what you did in your last comment?

I shall look at that tomorrow.